File: sdk-typescript/packages/podcast/src/podcast.ts
/**
 * WAVE SDK - Podcast API
 *
 * Podcast production, episode management, and distribution.
 */

import type { WaveClient, PaginationParams, PaginatedResponse, Timestamps } from '@wave-av/core';

export type EpisodeStatus = "draft" | "processing" | "published" | "scheduled" | "failed";
export type DistributionTarget = "spotify" | "apple" | "google" | "amazon" | "overcast";

export interface Podcast extends Timestamps {
  id: string;
  organization_id: string;
  title: string;
  description: string;
  cover_art_url?: string;
  rss_url?: string;
  category: string;
  language: string;
  explicit: boolean;
  author: string;
  email?: string;
  website?: string;
  subscriber_count: number;
  episode_count: number;
}
export interface Episode extends Timestamps {
  id: string;
  podcast_id: string;
  title: string;
  description: string;
  status: EpisodeStatus;
  audio_url?: string;
  duration_seconds: number;
  file_size_bytes: number;
  season_number?: number;
  episode_number?: number;
  published_at?: string;
  scheduled_at?: string;
  tags?: string[];
}
export interface PodcastAnalytics {
  podcast_id: string;
  total_downloads: number;
  unique_listeners: number;
  average_listen_duration: number;
  top_episodes: { episode_id: string; downloads: number }[];
  listener_geography: { country: string; count: number }[];
}
export interface PodcastDistribution {
  target: DistributionTarget;
  status: "connected" | "pending" | "error";
  url?: string;
}
export interface CreatePodcastRequest {
  title: string;
  description: string;
  category: string;
  language?: string;
  explicit?: boolean;
  author?: string;
  email?: string;
}
export interface CreateEpisodeRequest {
  podcast_id: string;
  title: string;
  description: string;
  audio_url?: string;
  season_number?: number;
  episode_number?: number;
  tags?: string[];
  scheduled_at?: string;
}

/**
 * Podcast production, episode management, RSS feeds, and distribution.
 *
 * @example
 * ```typescript
 * const podcast = await wave.podcast.create({ title: "My Show", description: "...", category: "Tech" });
 * await wave.podcast.createEpisode({ podcast_id: podcast.id, title: "Ep 1", description: "..." });
 * await wave.podcast.distribute(podcast.id, ["spotify", "apple"]);
 * ```
 */
export class PodcastAPI {
  private readonly client: WaveClient;
  private readonly basePath = "/v1/podcasts";
  constructor(client: WaveClient) {
    this.client = client;
  }

  async create(request: CreatePodcastRequest): Promise<Podcast> {
    return this.client.post<Podcast>(this.basePath, request);
  }
  async get(podcastId: string): Promise<Podcast> {
    return this.client.get<Podcast>(`${this.basePath}/${podcastId}`);
  }
  async update(podcastId: string, updates: Partial<CreatePodcastRequest>): Promise<Podcast> {
    return this.client.patch<Podcast>(`${this.basePath}/${podcastId}`, updates);
  }
  async remove(podcastId: string): Promise<void> {
    await this.client.delete(`${this.basePath}/${podcastId}`);
  }
  async list(params?: PaginationParams): Promise<PaginatedResponse<Podcast>> {
    return this.client.get<PaginatedResponse<Podcast>>(this.basePath, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }
  async createEpisode(request: CreateEpisodeRequest): Promise<Episode> {
    return this.client.post<Episode>(`${this.basePath}/${request.podcast_id}/episodes`, request);
  }
  async getEpisode(episodeId: string): Promise<Episode> {
    return this.client.get<Episode>(`/v1/episodes/${episodeId}`);
  }
  async updateEpisode(episodeId: string, updates: Partial<CreateEpisodeRequest>): Promise<Episode> {
    return this.client.patch<Episode>(`/v1/episodes/${episodeId}`, updates);
  }
  async removeEpisode(episodeId: string): Promise<void> {
    await this.client.delete(`/v1/episodes/${episodeId}`);
  }
  async publishEpisode(episodeId: string): Promise<Episode> {
    return this.client.post<Episode>(`/v1/episodes/${episodeId}/publish`);
  }
  async listEpisodes(
    podcastId: string,
    params?: PaginationParams,
  ): Promise<PaginatedResponse<Episode>> {
    return this.client.get<PaginatedResponse<Episode>>(`${this.basePath}/${podcastId}/episodes`, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }
  async getRSSFeed(podcastId: string): Promise<{ url: string; xml: string }> {
    return this.client.get<{ url: string; xml: string }>(`${this.basePath}/${podcastId}/rss`);
  }
  async getAnalytics(podcastId: string, params?: { period?: string }): Promise<PodcastAnalytics> {
    return this.client.get<PodcastAnalytics>(`${this.basePath}/${podcastId}/analytics`, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }
  async distribute(
    podcastId: string,
    targets: DistributionTarget[],
  ): Promise<PodcastDistribution[]> {
    return this.client.post<PodcastDistribution[]>(`${this.basePath}/${podcastId}/distribute`, {
      targets,
    });
  }
  async getDistributionStatus(podcastId: string): Promise<PodcastDistribution[]> {
    return this.client.get<PodcastDistribution[]>(`${this.basePath}/${podcastId}/distribution`);
  }
}

export function createPodcastAPI(client: WaveClient): PodcastAPI {
  return new PodcastAPI(client);
}

File: sdk-python/wave_av_sdk/podcast.py
"""WAVE SDK - Podcast API. Podcast production, episodes, and distribution."""
from __future__ import annotations
from typing import Any
from pydantic import BaseModel
from wave.client import WaveClient

class Podcast(BaseModel):
    id: str; organization_id: str; title: str; description: str; category: str; language: str = "en"; episode_count: int = 0; subscriber_count: int = 0; rss_url: str | None = None; created_at: str; updated_at: str

class Episode(BaseModel):
    id: str; podcast_id: str; title: str; description: str; status: str; audio_url: str | None = None; duration_seconds: float = 0; published_at: str | None = None; created_at: str; updated_at: str

class PodcastAPI:
    def __init__(self, client: WaveClient): self._client = client; self._base = "/v1/podcasts"
    def create(self, title: str, description: str, category: str, **kwargs: Any) -> Podcast: return Podcast(**self._client.post(self._base, json={"title": title, "description": description, "category": category, **kwargs}))
    def get(self, podcast_id: str) -> Podcast: return Podcast(**self._client.get(f"{self._base}/{podcast_id}"))
    def update(self, podcast_id: str, **kwargs: Any) -> Podcast: return Podcast(**self._client.patch(f"{self._base}/{podcast_id}", json=kwargs))
    def remove(self, podcast_id: str) -> None: self._client.delete(f"{self._base}/{podcast_id}")
    def list(self, **params: Any) -> dict: return self._client.get(self._base, params={k: v for k, v in params.items() if v is not None})
    def create_episode(self, podcast_id: str, title: str, description: str, **kwargs: Any) -> Episode: return Episode(**self._client.post(f"{self._base}/{podcast_id}/episodes", json={"title": title, "description": description, **kwargs}))
    def get_episode(self, episode_id: str) -> Episode: return Episode(**self._client.get(f"/v1/episodes/{episode_id}"))
    def publish_episode(self, episode_id: str) -> Episode: return Episode(**self._client.post(f"/v1/episodes/{episode_id}/publish"))
    def list_episodes(self, podcast_id: str, **params: Any) -> dict: return self._client.get(f"{self._base}/{podcast_id}/episodes", params={k: v for k, v in params.items() if v is not None})
    def get_rss_feed(self, podcast_id: str) -> dict: return self._client.get(f"{self._base}/{podcast_id}/rss")
    def get_analytics(self, podcast_id: str, **params: Any) -> dict: return self._client.get(f"{self._base}/{podcast_id}/analytics", params={k: v for k, v in params.items() if v is not None})
    def distribute(self, podcast_id: str, targets: list[str]) -> list[dict]: return self._client.post(f"{self._base}/{podcast_id}/distribute", json={"targets": targets})
